### ISSUE TITLE
Parsing of lines.txt to word_line.csv

### ISSUE THREAD

**Issue by ivo-1** [jc639/pytorch-handwritingCTC#5]
Hi,

I am trying to build a model that doesn't use CTC loss on the IAM (line) dataset and am in the process of creating the DataLoader. You have built a Dataset class and you use a .csv (namely `word_line.csv`) in order to create an instance of it if I understand your code correctly. It would be great if you could either push the code that parses the `lines.txt` (i.e. the solutions) or `word_line.csv` itself. I think it's not so trivial to achieve this and would be grateful if you could help me out.

Best,
ivo-1

**Comment by jc639** [comment]
I can't find my version of lines.txt or word_line.csv any more (I haven't looked at this in years to be honest - if you can post the first few lines of the file to remind me, I can definitely confirm) but from what i recall it is just a tab (\t) separated text file. You should be able to read it in with some variation of:

```
import pandas as pd

df = pd.read_csv('lines.txt', sep='\t')
df.to_csv('word_line.csv', index=False)
```

might need to play around with the `skiprows` parameter of pd.read_csv but I can't remember doing much more than that.

**Comment by ivo-1** [comment]
Thanks for the fast reply. Unfortunately, I don't think `lines.txt` is a tab separated file. Here are the first few lines (i.e. skiprows=22):

```
#--- lines.txt ---------------------------------------------------------------#
#
# iam database line information
#
# format: a01-000u-00 ok 154 19 408 746 1663 91 A|MOVE|to|stop|Mr.|Gaitskell|from
#
#     a01-000u-00  -> line id for form a01-000u
#     ok              -> result of word segmentation
#                            ok: line is correctly segmented
#                            err: segmentation of line has one or more errors
#
#                        notice: if the line could not be properly segmented
#                                the transcription and extraction of the whole
#                                line should not be affected negatively
#
#     154             -> graylevel to binarize line
#     19              -> number of components for this line
"#     408 746 1663 91 -> bounding box around this line in x,y,w,h format"
#
#     A|MOVE|to|stop|Mr.|Gaitskell|from
#                     -> transcription for this line. word tokens are separated
#                        by the character |
#
a01-000u-00 ok 154 19 408 746 1661 89 A|MOVE|to|stop|Mr.|Gaitskell|from
a01-000u-01 ok 156 19 395 932 1850 105 nominating|any|more|Labour|life|Peers
a01-000u-02 ok 157 16 408 1106 1986 105 is|to|be|made|at|a|meeting|of|Labour
a01-000u-03 err 156 23 430 1290 1883 70 M Ps|tomorrow|.|Mr.|Michael|Foot|has
a01-000u-04 ok 157 20 395 1474 1830 94 put|down|a|resolution|on|the|subject
```

Following that there is the task left of segmenting the transcription to the proper sentence, where simple search and replace does not suffice, as one doesn't want spaces before punctuation. But this is another issue that can be dealt with.

Looking forward to if you maybe did something else, now that you see the first lines of `lines.txt`. If not, then I'll just have to see how I deal with this problem of converting it into .csv. Thanks nonetheless!

**Comment by jc639** [comment]
Ah yeah this brings me back! I definitely didn't remember very well. I can't remember what I did exactly but I think it was something like:

```
import pandas as pd

fname = []
seg_ok = []
word = []
with open('lines.txt', 'r') as f:
   for line in f.readlines()[23:]: # might be from index 22 not sure whether first line is in file or not
       lsplit = line.split(' ')
       fname.append(lsplit[0])
       seg_ok.append(lsplit[1])
       word.append(lsplit[-1].replace('|', ' ').strip())

df = pd.DataFrame(data={'file': fname, 'ok': seg_ok, 'word': word})

# to remove the spaces before punctation you can use some regex - i didnt do this originally.
# Stackoverflow --> https://stackoverflow.com/a/18878970
df['word'] = df.word.str.replace(r'\s([?.!"](?:\s|$))', r'\1', regex=True)

df.to_csv('word_line.csv', index=False)
print(df)
```
With the snippet of lines.txt I get this:
![Screenshot 2021-09-14 at 16 50 08](https://user-images.githubusercontent.com/28015710/133291191-ee6116fe-a7ab-4739-b1fa-55629ba6575d.png)



Hope that helps.

**Comment by ivo-1** [comment]
Hi, that's fantastic! Thanks so much, it really helps me out and solves the issue I had.
